compiler/optimizer/abstractinterpreter/InliningMethodSummary.cpp

```diff
@@ -0,0 +1,155 @@
+      {
+      TR::PotentialOptimizationPredicate* predicate = (*_optsByArg[argPos])[i];
+      if (predicate->test(arg))
+         {
+         benefit += 1;
+         }
+      }
+
+   return benefit;
+   }
+
+void TR::InliningMethodSummary::trace(TR::Compilation* comp)
+   {
+   traceMsg(comp, "Inlining Method Summary:\n");
+
+   if (_optsByArg.size() == 0)
+      {
+      traceMsg(comp, "EMPTY\n\n");
+      return;
+      }
+
+   for (size_t i = 0; i < _optsByArg.size(); i ++) 
+      {
+      if (_optsByArg[i] != NULL)
+         {
+         for (size_t j = 0; j < _optsByArg[i]->size(); j ++)
+            {
+            TR::PotentialOptimizationPredicate* predicate = (*_optsByArg[i])[j];
+
+            traceMsg(comp, "%s @%d for Argument %d ", predicate->getName(), predicate->getBytecodeIndex(), i);
+            predicate->trace(comp);
+            traceMsg(comp, "\n");
+            }
+         }
+      }
+
+   }
+
+
+void TR::InliningMethodSummary::addPotentialOptimizationByArgument(TR::PotentialOptimizationPredicate* predicate, uint32_t argPos)
+   {
+   if (_optsByArg.size() <= argPos)
+      _optsByArg.resize(argPos + 1);
+      
+   if (!_optsByArg[argPos])
+       _optsByArg[argPos] = new (region()) PredicateContainer(region());
+      
+   _optsByArg[argPos]->push_back(predicate);
+   }
+
+const char* TR::PotentialOptimizationPredicate::getName()
+   {
+   switch (_kind)
+      {
+      case Kind::BranchFolding:
+         return "Branch Folding";
+      case Kind::CheckCastFolding:
+         return "CheckCast Folding";
+      case Kind::NullCheckFolding:
+         return "NullCheck Folding";
+      case Kind::InstanceOfFolding:
+         return "InstanceOf Folding";
+      default:
+         TR_ASSERT_FATAL(false, "Unexpected Kind");
+      }
+   }
+
+bool TR::PotentialOptimizationVPPredicate::holdPartialOrderRelation(TR::VPConstraint* valueConstraint, TR::VPConstraint* testConstraint)  
+   {
+   if (testConstraint->asIntConstraint()) // partial relation for int constraint
+      {
+      if (testConstraint->getLowInt() <= valueConstraint->getLowInt() && testConstraint->getHighInt() >= valueConstraint->getHighInt()) 
+         return true;
+      else 
+         return false;
+      }
+   else if (testConstraint->asClassPresence()) // partial relation for nullness
+      {
+      if (testConstraint->isNonNullObject() && valueConstraint->isNonNullObject())
+         return true;
+      else if (testConstraint->isNullObject() && valueConstraint->isNullObject())
+         return true;
+      else 
+         return false;
+      }
+   else if (testConstraint->asClassType()) // partial relation for class types
+      {
+      if (valueConstraint->isNonNullObject() && valueConstraint->getClass())
+         {
+         TR_YesNoMaybe yesNoMaybe = _vp->fe()->isInstanceOf(valueConstraint->getClass(), testConstraint->getClass(), valueConstraint->isFixedClass(), true);
+
+         if (yesNoMaybe != TR_maybe)
+            return true;
```

Comment:
Technically, this is doable, however, it may require changes to the logic of VPConstraint since our predicates are based on the VPConstraint.  We may not give long-term support related to the changes if we decide to do so.  Thus, we may not do this at this moment

Reply:
I don't think I see why that would require modifying `VPConstraint`. It would be two cases external to `VPConstraint`:
1. Here is a `VPConstraint` and an optimization is possible whenever the parameter value can be guaranteed to satisfy it
2. Here is a `VPConstraint` and an optimization is possible whenever the parameter value can be guaranteed _not_ to satisfy it

Reply:
I suppose the case where `checkcast` is guaranteed to fail (non-null and not type `T`) is only expressible this way if there is a representation for negation within `VPConstraint`. To make that expressible without modifying `VPConstraint`, we would need the potential optimization to be a conjunction of clauses of form (1) and (2) instead of just a single such clause

Reply:
Reflecting our discussion here: We agreed to separate the `instanceof`/`checkcast` transformations into separate representations as described above (cases (1) and (2)).

Given my previous comment, we also agreed for the moment to disregard the failing `checkcast` case. **However**, upon further reflection, I'm now going to contradict my previous comment. I think the failing `checkcast` case really is expressible simply as the negation (i.e. case (2)) of a `VPClassType`: "not type `T`". Because `VPClassType` includes null, a possibly-null reference to an unrelated type `U` is *not* known not to satisfy the constraint - in particular, the value at runtime will satisfy the constraint whenever it happens to be null. OTOH, a *non-null* reference to `U` *is* known not to satisfy the constraint. In other words, "not type `T`" implies non-null, so the "non-null and" part of the condition is unnecessary

Reply:
I should mention as well that on this latest understanding of the failing `checkcast` condition, we have a nice symmetry: for each of `instanceof`/`checkcast`, there is a predicate *P* for success. In each case, *P* can be expressed precisely as a VP constraint:
- for `instanceof`, *P* is "non-null and type `T`", and
- for `checkcast`, *P* is just "type `T`"

These are precise characterizations, so if we can prove *P*, we get guaranteed success (true/fallthrough), and if we can prove *not-P*, we get guaranteed failure (false/throw)